### PR TITLE
KAFKA-6945: KIP-373, allow users to create delegation token for others

### PR DESCRIPTION
**Pull request by viktorsomogyi**
This PR adds the capability of users to create delegation token to other users.

KIP-373: https://cwiki.apache.org/confluence/display/KAFKA/KIP-373%3A+Allow+users+to+create+delegation+tokens+for+other+users

In this use case, a superuser with username ‘superuser’ wants to run kafka clients on behalf of a user 'joe'. The 'superuser' has secure authentication credentials (kerberos, SSL, SCRAM) but user 'joe' doesn’t have any. The clients are required to run as user 'joe' and authorizations are required to be done as user 'joe.' In this case, 'superuser' can get a delegation token for user 'joe', and use the generated token to run the Kafka clients. This will mimic the impersonation functionality. This will help the stream processing frameworks/libs (Apache Spark, Storm, Kafka Streams) to run the jobs (Kafka clients) as submitted users.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by viktorsomogyi** [comment]
Finished rebasing and coding (and therefore force pushing), I'll switch the pr into "ready for review" state now.

**Comment by viktorsomogyi** [comment]
@omkreddy @rajinisivaram would you please help in the review to get this long outstanding KIP done?

**Comment by sachmo99** [comment]
Hi @viktorsomogyi ! Any update on this KIP implementation plan?

**Comment by viktorsomogyi** [comment]
@sachmo99 I pinged @rajinisivaram and @omkreddy in email, hopefully they respond and I'd be happy to rebase this change on top of trunk and push it in.

**Comment by omkreddy** [comment]
Sorry for missing this PR. I will help to review and merge the PR .

**Comment by viktorsomogyi** [comment]
Thank you Manikumar!
 I'll rebase the change then, will ping you when I'm done (since it is late Friday here it might be on Monday).

**Comment by viktorsomogyi** [comment]
@omkreddy I'm finished with rebasing the PR. There were a couple of test failures that I needed to fix but as I see the currently failing ones are flaky.

File: core/src/main/scala/kafka/admin/AclCommand.scala [hunk 1]
```diff
@@ -467,8 +467,11 @@ object AclCommand extends Logging {
     if (opts.options.has(opts.delegationTokenOpt))
       opts.options.valuesOf(opts.delegationTokenOpt).forEach(token => resourceFilters += new ResourcePatternFilter(JResourceType.DELEGATION_TOKEN, token.trim, patternType))
 
+    if (opts.options.has(opts.userPrincipalOpt))
+      resourceFilters += new ResourcePatternFilter(JResourceType.USER, opts.options.valueOf(opts.userPrincipalOpt), patternType)
+
     if (resourceFilters.isEmpty && dieIfNoResourceFound)
-      CommandLineUtils.printUsageAndDie(opts.parser, "You must provide at least one resource: --topic <topic> or --cluster or --group <group> or --delegation-token <Delegation Token ID>")
```
Comment (omkreddy) [inline core/src/main/scala/kafka/admin/AclCommand.scala:471]: nit: unwanted space

File: core/src/test/scala/integration/kafka/api/DelegationTokenEndToEndAuthorizationWithOwnerTest.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ */
+package kafka.api
+
+import kafka.admin.AclCommand
+import kafka.utils.JaasTestUtils
+import org.apache.kafka.clients.admin.{Admin, CreateDelegationTokenOptions}
+import org.apache.kafka.common.security.auth.KafkaPrincipal
+import org.apache.kafka.common.security.token.delegation.DelegationToken
+
+class DelegationTokenEndToEndAuthorizationWithOwnerTest extends DelegationTokenEndToEndAuthorizationTest {
```
Comment (omkreddy) [inline core/src/test/scala/integration/kafka/api/DelegationTokenEndToEndAuthorizationWithOwnerTest.scala:31]: Can we add a negative test (i.e) try to create token for a user without valid permissions?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 3]
```diff
@@ -2926,24 +2926,36 @@ class KafkaApis(val requestChannel: RequestChannel,
...
 
+    val ownerPrincipalName = createTokenRequest.data().ownerPrincipalName()
+    val owner = if (ownerPrincipalName == null || ownerPrincipalName.isEmpty) {
+      request.context.principal
+    } else {
+      new KafkaPrincipal(createTokenRequest.data().ownerPrincipalType(), ownerPrincipalName)
+    }
+    val requester = request.context.principal
+
     if (!allowTokenRequests(request))
```
Comment (omkreddy) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: We should also add authorizer check if `owner != requester` ? also a test for this scenario.

File: core/src/main/scala/kafka/admin/AclCommand.scala [hunk 4]
```diff
@@ -467,8 +467,11 @@ object AclCommand extends Logging {
     if (opts.options.has(opts.delegationTokenOpt))
       opts.options.valuesOf(opts.delegationTokenOpt).forEach(token => resourceFilters += new ResourcePatternFilter(JResourceType.DELEGATION_TOKEN, token.trim, patternType))
 
+    if (opts.options.has(opts.userPrincipalOpt))
```
Comment (omkreddy) [inline core/src/main/scala/kafka/admin/AclCommand.scala:470]: Should we update AclCommandTest ? Can we add any checks to checkArgs() method.

**Review by omkreddy — COMMENTED** [review]
@viktorsomogyi Thanks for the PR. left minor comments. Can we run the test again

Can we also update delegation token/acl related docs.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 3 above)
Comment (viktorsomogyi) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I'll solve it like this (and added tests too):
```
    if (!allowTokenRequests(request)) {
      requestHelper.sendResponseMaybeThrottle(request, requestThrottleMs =>
        CreateDelegationTokenResponse.prepareResponse(request.context.requestVersion, requestThrottleMs,
          Errors.DELEGATION_TOKEN_REQUEST_NOT_ALLOWED, owner, requester))
    } else if (!owner.equals(requester) && !authHelper.authorize(request.context, CREATE_TOKENS, USER, owner.toString)) {
      requestHelper.sendResponseMaybeThrottle(request, requestThrottleMs =>
        CreateDelegationTokenResponse.prepareResponse(request.context.requestVersion, requestThrottleMs,
          Errors.DELEGATION_TOKEN_AUTHORIZATION_FAILED, owner, requester))
    } else {
```

**Comment by viktorsomogyi** [comment]
retest this please

**Comment by viktorsomogyi** [comment]
@omkreddy I've addressed your comments. Would you please check the PR again?
Also I see that some of the configurations fail to build but I don't know why is that. Locally they're fine.

File: core/src/test/scala/unit/kafka/security/token/delegation/DelegationTokenManagerTest.scala [hunk 5]
```diff
@@ -315,13 +314,20 @@ class DelegationTokenManagerTest extends QuorumTestHarness  {
...
+      def authorizeToken(tokenId: String): Boolean = {
         val requestContext = AuthorizerUtils.sessionToRequestContext(hostSession)
         val action = new Action(AclOperation.DESCRIBE,
           new ResourcePattern(DELEGATION_TOKEN, tokenId, LITERAL), 1, true, true)
         aclAuthorizer.authorize(requestContext, List(action).asJava).asScala.head == AuthorizationResult.ALLOWED
       }
-      def eligible(token: TokenInformation) = DelegationTokenManager.filterToken(requestPrincipal, Option(requestedOwners), token, authorizeToken)
+      def authorizeRequester(owner: KafkaPrincipal): Boolean = {
+        val requestContext = AuthorizerUtils.sessionToRequestContext(hostSession)
+        val action = new Action(AclOperation.DESCRIBE,
```
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/security/token/delegation/DelegationTokenManagerTest.scala]: Do we have a test to exercise this method alone?  why are we using `DESCRIBE` Operation here?

File: core/src/main/scala/kafka/security/authorizer/AclEntry.scala [hunk 6]
```diff
@@ -100,9 +100,10 @@ object AclEntry {
     resourceType match {
       case ResourceType.TOPIC => Set(READ, WRITE, CREATE, DESCRIBE, DELETE, ALTER, DESCRIBE_CONFIGS, ALTER_CONFIGS)
       case ResourceType.GROUP => Set(READ, DESCRIBE, DELETE)
-      case ResourceType.CLUSTER => Set(CREATE, CLUSTER_ACTION, DESCRIBE_CONFIGS, ALTER_CONFIGS, IDEMPOTENT_WRITE, ALTER, DESCRIBE)
+      case ResourceType.CLUSTER => Set(CREATE, CLUSTER_ACTION, DESCRIBE_CONFIGS, ALTER_CONFIGS, IDEMPOTENT_WRITE, ALTER, DESCRIBE, CREATE_TOKENS, DESCRIBE_TOKENS)
```
Comment (omkreddy) [inline core/src/main/scala/kafka/security/authorizer/AclEntry.scala]: Do we need this? We have not defined these operations in KIP https://cwiki.apache.org/confluence/display/KAFKA/KIP-373%3A+Allow+users+to+create+delegation+tokens+for+other+users#KIP373:Allowuserstocreatedelegationtokensforotherusers-Operations

**Review by omkreddy — COMMENTED** [review]
@viktorsomogyi Thanks for the updates. We are close to merge the PR. left couple of minor comments.

File: core/src/main/scala/kafka/security/authorizer/AclEntry.scala (hunk 6 above)
Comment (viktorsomogyi) [reply core/src/main/scala/kafka/security/authorizer/AclEntry.scala]: Oh, thanks for pointing it out, I think it wasn't intentional. Will remove it.

File: core/src/test/scala/unit/kafka/security/token/delegation/DelegationTokenManagerTest.scala (hunk 5 above)
Comment (viktorsomogyi) [reply core/src/test/scala/unit/kafka/security/token/delegation/DelegationTokenManagerTest.scala]: Good point, should be `DESCRIBE_TOKENS`. I'll correct this. I believe the `DelegationTokenEndToEndAuthorizationWithOwnerTest` tests the relevant code path (that uses `DelegationTokenManager.filterToken`) but that uses the correct permissions (it sets them through the ACL command). Will get back to you with the test for this.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 7]
```diff
@@ -3285,7 +3285,7 @@ class KafkaApisTest {
 
     assertEquals(metadataCache.getControllerId.get, describeClusterResponse.data.controllerId)
     assertEquals(clusterId, describeClusterResponse.data.clusterId)
-    assertEquals(8096, describeClusterResponse.data.clusterAuthorizedOperations)
+    assertEquals(32672, describeClusterResponse.data.clusterAuthorizedOperations)
```
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: this change is not required now

File: core/src/test/scala/unit/kafka/security/token/delegation/DelegationTokenManagerTest.scala (hunk 5 above)
Comment (omkreddy) [reply core/src/test/scala/unit/kafka/security/token/delegation/DelegationTokenManagerTest.scala]: This is still using `DESCRIBE` operation

**Comment by viktorsomogyi** [comment]
@omkreddy would you please check it once you get some time?

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java [hunk 8]
```diff
@@ -364,7 +365,8 @@ public enum Errors {
     INCONSISTENT_TOPIC_ID(103, "The log's topic ID did not match the topic ID in the request", InconsistentTopicIdException::new),
     INCONSISTENT_CLUSTER_ID(104, "The clusterId in the request does not match that found on the server", InconsistentClusterIdException::new),
     TRANSACTIONAL_ID_NOT_FOUND(105, "The transactionalId could not be found", TransactionalIdNotFoundException::new),
-    FETCH_SESSION_TOPIC_ID_ERROR(106, "The fetch session encountered inconsistent topic ID usage", FetchSessionTopicIdException::new);
+    FETCH_SESSION_TOPIC_ID_ERROR(106, "The fetch session encountered inconsistent topic ID usage", FetchSessionTopicIdException::new),
+    USER_AUTHORIATION_FAILED(107, "User authorization failed", UserAuthorizationException::new);
```
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: minor typo: `USER_AUTHORIATION_FAILED` => `USER_AUTHORIAZATION_FAILED`

Comment (omkreddy) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: Also we are not using this error. Can we remove this. We can just `DELEGATION_TOKEN_AUTHORIZATION_FAILED` error for all token related authorisation issues.

**Review by omkreddy — APPROVED** [review]
@viktorsomogyi Thanks for the PR. LGTM.

Can you also update docs for the updates.

**Comment by viktorsomogyi** [comment]
@omkreddy thanks, I'll open a PR likely tomorrow for the docs.

**Pull request merged**